Attempting to fix missing symbols in stack trace

hkaiser opened the pull request:
Fixes #4580

This also adds a runtime configuration option to specify the stack trace depth: `--hpx:ini=hpx.trace_depth=<N>`.

Note: the symbol lookup fix proposed here was not actually tested. @teonnik, @msimberg would you be able to verify this, please?

msimberg commented:
Thanks @hkaiser for looking into this! I'll give this a try.

teonnik commented:
I just tried out the branch but, unfortunately, symbols are still missing:

```
9 frames:
0x7fdc5765b2cf  : hpx::util::stack_trace::trace(void**, unsigned long) + 0x1f in /home/teonnik/software/spack/opt/spack/linux-arch-broadwell/clang-9.0.1/hpx-backtrace-3eqcxqd73xxx5ljxvoi7na35s7pbo6zt/lib/libhpxd.so.1
0x564a6717befb  : hpx::util::backtrace::backtrace(unsigned long) + 0x9b in ./build/backtrace
0x564a6717bd7a  : ./build/backtrace(+0x15d7a) [0x564a6717bd7a] + 0x564a6717bd7a in ./build/backtrace
0x564a6717be29  : ./build/backtrace(+0x15e29) [0x564a6717be29] + 0x564a6717be29 in ./build/backtrace
0x564a6717be39  : ./build/backtrace(+0x15e39) [0x564a6717be39] + 0x564a6717be39 in ./build/backtrace
0x564a6717be54  : ./build/backtrace(+0x15e54) [0x564a6717be54] + 0x564a6717be54 in ./build/backtrace
0x564a6717e83f  : ./build/backtrace(+0x1883f) [0x564a6717e83f] + 0x564a6717e83f in ./build/backtrace
0x7fdc54f70023  : __libc_start_main + 0xf3 in /usr/lib/libc.so.6
0x564a6717bc7e  : ./build/backtrace(+0x15c7e) [0x564a6717bc7e] + 0x564a6717bc7e in ./build/backtrace
```

I also tried `--hpx:ini=hpx.trace_depth=3` but it always prints 9 frames.

hkaiser commented:
> I just tried out the branch but, unfortunately, symbols are still missing:
> 9 frames:
> 0x7fdc5765b2cf  : hpx::util::stack_trace::trace(void**, unsigned long) + 0x1f in /home/teonnik/software/spack/opt/spack/linux-arch-broadwell/clang-9.0.1/hpx-backtrace-3eqcxqd73xxx5ljxvoi7na35s7pbo6zt/lib/libhpxd.so.1
> 0x564a6717befb  : hpx::util::backtrace::backtrace(unsigned long) + 0x9b in ./build/backtrace
> 0x564a6717bd7a  : ./build/backtrace(+0x15d7a) [0x564a6717bd7a] + 0x564a6717bd7a in ./build/backtrace
> 0x564a6717be29  : ./build/backtrace(+0x15e29) [0x564a6717be29] + 0x564a6717be29 in ./build/backtrace
> 0x564a6717be39  : ./build/backtrace(+0x15e39) [0x564a6717be39] + 0x564a6717be39 in ./build/backtrace
> 0x564a6717be54  : ./build/backtrace(+0x15e54) [0x564a6717be54] + 0x564a6717be54 in ./build/backtrace
> 0x564a6717e83f  : ./build/backtrace(+0x1883f) [0x564a6717e83f] + 0x564a6717e83f in ./build/backtrace
> 0x7fdc54f70023  : __libc_start_main + 0xf3 in /usr/lib/libc.so.6
> 0x564a6717bc7e  : ./build/backtrace(+0x15c7e) [0x564a6717bc7e] + 0x564a6717bc7e in ./build/backtrace
> 
> I also tried --hpx:ini=hpx.trace_depth=3 but it always prints the same 9 frames.

Thanks for checking! We have some improvement, however. The backtrace does not show `'???'` for the symbol names anymore but is slightly more descriptive.

teonnik commented:
> I also tried --hpx:ini=hpx.trace_depth=3 but it always prints the same 9 frames.

Oops, that was because the I had the value manually specified in the example `hpx::util::backtrace(10)`. 

I changed the code to invoke `hpx::init()` and throw a `HPX_THROW_EXCEPTION(hpx::no_success, "raise_exception", "simulated error");` instead, the parameter seems not to have an effect as 20 frames are printed.

msimberg commented on src/runtime.cpp:
```diff
@@ -84,7 +84,11 @@ namespace hpx
+                    util::from_string<int>(get_config_entry(
+                        "hpx.trace_depth", HPX_HAVE_THREAD_BACKTRACE_DEPTH));
+                std::cerr << "{stack-trace}: " << hpx::util::trace(trace_depth)
+                          << "\n";
```
I think this change needs to be applied in `custom_exception_info.cpp` as well to have it used for exceptions.

hkaiser replied:
Yep, I just applied this change there as well. Things should be better now.

hkaiser commented:
@teonnik After investigating and reading about this some more, I have added the `-Wl,-E` option while linking executables (more generally while linking with hpx_init, not sure if this is what we want...) to force all symbols being added to the executables dynamic symbol table. This should make the symbols visible to `dladdr`. Please try again.

teonnik commented:
@hkaiser The symbols are now resolved (even without calling hpx::init()). Thank you!

```
➜ ./build/backtrace
10 frames:
0x7fee9b3298ef  : hpx::util::stack_trace::trace(void**, unsigned long) [0x1f] in /home/teonnik/software/spack/opt/spack/linux-arch-broadwell/clang-9.0.1/hpx-backtrace-3eqcxqd73xxx5ljxvoi7na35s7pbo6zt/lib/libhpxd.so.1
0x560c45cd4d5b  : hpx::util::backtrace::backtrace(unsigned long) [0x9b] in ./build/backtrace
0x560c45cd419a  : func_1() [0x2a] in ./build/backtrace
0x560c45cd4249  : func_2() [0x9] in ./build/backtrace
0x560c45cd4259  : func_3() [0x9] in ./build/backtrace
0x560c45cd4274  : test_hpx(int, char**) [0x14] in ./build/backtrace
0x560c45cd42a2  : main [0x22] in ./build/backtrace
0x560c45cdbe2f  : __wrap_main [0x3f] in ./build/backtrace
0x7fee98c97023  : __libc_start_main [0xf3] in /usr/lib/libc.so.6
0x560c45cd409e  : _start [0x2e] in ./build/backtrace
```

The command line argument `--hpx:ini=hpx.trace_depth=` also works with HPX exceptions, which is great!

Would it be possible to omit the first 2 frames as they belong to the `backtrace` utility itself? :

```
0x7fee9b3298ef  : hpx::util::stack_trace::trace(void**, unsigned long) [0x1f] in /home/teonnik/software/spack/opt/spack/linux-arch-broadwell/clang-9.0.1/hpx-backtrace-3eqcxqd73xxx5ljxvoi7na35s7pbo6zt/lib/libhpxd.so.1
0x560c45cd4d5b  : hpx::util::backtrace::backtrace(unsigned long) [0x9b] in ./build/backtrace
```

hkaiser commented:
@teonnik I have removed the first two stack frames from the listing. Please note that I have added the `HPX_WITH_STACKTRACES_STATIC_SYMBOLS` configuration option that enables the `-Wl,-E` linker option as some of our tests started to segfault (see: https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/3278/workflows/2f7de056-d7fe-4ea9-a050-d582aa9d264f/jobs/186217/steps). I don't know why this is happening, however. This new option defaults to `OFF`, which means you will have to explicitly enable it.

hkaiser commented:
> @teonnik I have removed the first two stack frames from the listing. Please note that I have added the HPX_WITH_STACKTRACES_STATIC_SYMBOLS configuration option that enables the -Wl,-E linker option as some of our tests started to segfault (see: https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/3278/workflows/2f7de056-d7fe-4ea9-a050-d582aa9d264f/jobs/186217/steps). I don't know why this is happening, however. This new option defaults to OFF, which means you will have to explicitly enable it.

Further investigation shows that the segfault happens in libdl while attempting to collect a stack back trace. I'm inclined to declare this as 'not our problem' at this point.
